### ISSUE TITLE
Support Conditional Writes in OVHcloud Object Storage to Prevent Overwrites

### ISSUE THREAD

**Issue by mmiglier** [ovh/public-cloud-roadmap#671]
### Description:

As an OVHcloud customer using **Object Storage**,  
I want **support for conditional writes (e.g., using the `If-None-Match` header) to prevent overwriting existing objects**,  
so that **I can ensure data integrity and enforce immutability in workflows**.

---

### Use Case:

Attempting a conditional write currently results in the error:  
`An error occurred (NotImplemented) when calling the PutObject operation: Conditional object PUTs are not supported.`  

This feature would simplify enforcing non-overwrite policies directly in OVHcloud Object Storage, without relying on external application checks.

---

### Proposal:

Enable conditional writes via the `If-None-Match` header and allow enforcement through IAM policies.  

---

### Benefits:

- Ensures data integrity by preventing accidental overwrites.  
- Reduces complexity by eliminating the need for application-layer checks.  
- Aligns with industry standards and S3-compatible workflows.  

---

This issue aligns with common data governance and compliance needs.

**Comment by lason-ovh** [comment]
Hi @mmiglier ! Thanks, for the request. We will gladly look into it.

However, please note that we also have the Versioning and Object lock feature that help prevent accidental deletes and overwrites:
https://help.ovhcloud.com/csm/en-ie-public-cloud-storage-s3-versioning?id=kb_article_view&sysparm_article=KB0063858
https://help.ovhcloud.com/csm/en-ie-public-cloud-storage-s3-managing-object-lock?id=kb_article_view&sysparm_article=KB0047398

Are these features not enough to answer your need?

**Comment by mmiglier** [comment]
Yes, actually, object locking should be enough for my use case thanks

**Comment by olivierboudet** [comment]
Conditional Writes is nedded to support new terraform state locking if we want to store state in ovh bucket.

**Comment by Libr4rian** [comment]
It looks like I'm having the same problem trying to use an s3 bucket as a terraform backend with use_lockfile set to true.

**Comment by lason-ovh** [comment]
Thanks for the replies, we will look into it. To be honest, we are planning to work on major features such as bucket policies, bucket observability metrics and event notififactions in the forseeable future. If conditional writes become highly requested, we might reconsider our priorities.
By the way, we are also considering releasing our own TF provider to fully support OVH object storage.

**Comment by tonyalaribe** [comment]
I ran into this limitation while trying to use ovh as an aws S3-compatible storage for deltalake. 
It's already supported by most tools like minio and others.

**Comment by damz** [comment]
We would also love to see preconditions supported on OVHcloud Object Storage.

Namely, supporting:

* `If-Match: <Etag>` and `If-None-Match: "*"` on `CompleteMultipartUpload`, `PutObject`, and `CopyObject`; as well as
* `If-Match: <Etag>` on `DeleteObject`

At this point, all three major object storage systems (AWS S3, Google Storage and Azure Blob Storage) support preconditions (in sightly different forms, but they all support preconditions on writes and deletions), and we are using these as the foundation of our next generation storage.

In addition, we are planning on using preconditions as a way to migrate from the OVHcloud swift-based object storage to the new S3-compatible version.

**Comment by pushrax** [comment]
We had some customers asking if they can run our vector database on OVHcloud; this feature is blocking it. https://turbopuffer.com/

**Comment by yemaney** [comment]
This is blocking our team from using OVH cloud storage object storage for iceberg.

**Comment by ockhamlabs** [comment]
@lason-ovh this is a breaking issue for us. We are unable to use this for our data lake to work with query engines while its working with other cloud providers. We would appreciate expedited help on this

**Comment by lason-ovh** [comment]
> [@lason-ovh](https://github.com/lason-ovh) this is a breaking issue for us. We are unable to use this for our data lake to work with query engines while its working with other cloud providers. We would appreciate expedited help on this

Noted! we will look into it and see if it fits in our roadmap for Q3 2025. Can't make promises yet though

**Comment by lason-ovh** [comment]
Hello all, in order to help us prioritize and be able to deliver value quickly, we would need your feedbacks on this feature.
As a reminder, conditionnal writes are achieved via the following http headers:

|| operation || header ||
| put-object | if-match / if-none-match |
| copy-object | x-amz-copy-source-if-match / x-amz-copy-source-if-none-match <br> x-amz-copy-source-if-modified-since / x-amz-copy-source-if-unmodified-since |
| delete-object |  if-match |
| complete-multipart-upload | if-match / if-none-match |

**Comment by lason-ovh** [comment]
Hello all, in order to help us prioritize and be able to deliver value quickly, we would need your feedbacks on this feature.
As a reminder, conditionnal requests are achieved via the following http headers:

| operation | header |
| ----------- | -------- | 
| put-object | if-match <br> if-none-match |
| copy-object | x-amz-copy-source-if-match <br> x-amz-copy-source-if-none-match <br> x-amz-copy-source-if-modified-since <br> x-amz-copy-source-if-unmodified-since |
| delete-object |  if-match |
| complete-multipart-upload | if-match <br> if-none-match |
| get-object | if-match <br> if-none-match <br> if-modified-since <br> if-unmodified-since |
| head-object | if-match <br> if-none-match <br> if-modified-since <br> if-unmodified-since |

Among thoses operations and depending on your use case, which are the ones that are critical?
Would it be correct to suggest the following prioritization?

| release | header | supported operation | comment |
| -------- | -------- |  --------- | ---------- |
|  1 | if-match <br> if-none-match | PutObject<br>DeleteObject <br> GetObject <br> HeadObject <br> CompleteMultipartUpload* | *a specific case will not be supported i.e the case when user deletes an object with the same key that did not exist at the moment of the CreateMultipartUpload operation |
| 2 | x-amz-copy-source-if-match <br> x-amz-copy-source-if-none-match <br> x-amz-copy-source-if-modified-since <br> x-amz-copy-source-if-unmodified-since | CopyObject | |
| 3 | if-match <br> if-none-match | CompleteMultipartUpload* | *technical limitations that are currently being addressed prevent us from delivering the support for CompleteMultipartUpload operation in the medium-term in the case user deletes an object that did not exist at the moment of the CreateMultipartUpload |
| 4 | if-modified-since <br> if-unmodified-since | GetObject <br> HeadObject | |

Let us know what you think :)

**Comment by damz** [comment]
On our side, supporting these on `CompleteMultipartUpload` is essential.

The minimum viable support for us is:

* `If-Match: <Etag>` and `If-None-Match: "*"` on `CompleteMultipartUpload`; as well as
* `If-Match: <Etag>` on `DeleteObject`

**Comment by jsantist3ban** [comment]
Any news on that? I'm also interested in this feature.

**Comment by lason-ovh** [comment]
> Any news on this? I'm also interested in this feature.

we are currently in the design phase on this in order to be able to plan it in our roadmap, we will let you know when we get a reliable ETA

**Comment by AntoineRoll** [comment]
Hello, just to fuel the interest feedback, I've ran into an issue because of this unsupported feature through the usage of the deltalake library: https://github.com/delta-io/delta-rs/issues/3925

Looking forward to have this in prod :)

**Comment by agoude** [comment]
Thank you @AntoineRoll for the feedback and I understand the need. We'll share more soon about it as we are re prioritizing topics for the upcoming quarter.

**Comment by ciocan** [comment]
latest version of elasticsearch (9) also require conditional writes for data backups.

**Comment by leela782645** [comment]
We also need this feature as we want to use OVH Object Storage for Kafka Backups via Kafka Connect and the LensesIO StreamReactor S3 Sink which is expecting conditional writes.

**Comment by ganesh-rao** [comment]
The Milvus vector database also does not work on OVH Object Storage.

`panic: CheckIfConditionWriteSupport failed: PutObjectIfNoneMatch not supported or failed.`

**Comment by napierala** [comment]
Any news on this feature? The proposed release timeline is perfect for our use case.

**Comment by solracsf** [comment]
Conditional writes are part of the core S3 spec.
Providers that don't support them aren't S3-compatible, regardless of how they market themselves.

**Comment by agoude** [comment]
Hi all, we know many of you are eager to see this feature live, it’s been one of the most requested ones, and we completely understand why it matters. We had to shift priorities recently to focus on a few critical areas, which delayed progress here.
That said, it is still obviously on our radar and actively being worked on. No firm date yet, but we hope to share good news soon. Thanks a lot for your patience and for keeping the discussion going, it really helps us focus on what matters most 🙏 !

**Comment by Barre** [comment]
I've been asked by many OVH customers about this. Lack of support prevents [ZeroFS](https://github.com/Barre/ZeroFS) from working with OVH's object store. 

Eg. https://github.com/Barre/ZeroFS/discussions/335

**Comment by theodore-evans** [comment]
Please push on this! conditional PUT (`If-None-Match: *` on `PutObject`) is a core S3 feature that deltalake, iceberg, Terraform, etc. all depend on. AWS, GCS, and Azure have supported it for years. Without it, OVH Object Storage is incompatible with the modern data lakehouse ecosystem.

OVH has a really strong headstart in the European sovereign cloud market, but gaps like this push teams back toward the US hyperscalers, or whichever european cloud provider offers it first...

EDIT: looks like Exoscale does https://community.exoscale.com/product/storage/object-storage/how-to/conditionalrequests/

**Comment by dekiesel** [comment]
This feature would increase compatibility with proxmox backup server, which would make ovh the ideal backend for s3-hosted proxmox backups of VMs and LXCs

**Comment by loic-bellinger** [comment]
+1 from me too — running into the exact same limitation with delta-rs / Delta Lake tables on OVH Object Storage.

**Comment by lason-ovh** [comment]
Hi @all, just so you know, we are currently working on this feature, ETA is end of May 2026. We'll let you know as soon as we release it

**Comment by lason-ovh** [comment]
Hey all,
This is the updated release schedule for this feature:

| release | header | supported operation | ETA|
| -------- | -------- |  --------- | ---------- |
|  1 | if-match <br> if-none-match | PutObject<br>DeleteObject <br> GetObject <br> HeadObject <br> CompleteMultipartUpload | W20 |
| 2 | if-match <br> if-none-match | support for objects encrypted with SSE-C (customer managed encryption key) | W23 |
| 3 | x-amz-copy-source-if-match <br> x-amz-copy-source-if-none-match <br> x-amz-copy-source-if-modified-since <br> x-amz-copy-source-if-unmodified-since | CopyObject | not planned yet|

For the 3rd release (CopyObject) , let us know if it is a use case you want us to implement by liking this comment

**Comment by LePiegro** [comment]
Hello, any update on release 1?

**Comment by lason-ovh** [comment]
> Hello, any update on release 1?

Hi, release 1 is to be released next week ;) We'll let you know as soon as it is available

**Comment by loic-bellinger** [comment]
Any update on release 1 ?

**Comment by kofshikz** [comment]
Happy W23 everybody, hope this week we will get releases 1 and 2 together!

**Comment by agoude** [comment]
@loic-bellinger & @kofshikz and all, apologies for the delay. The roll-out ran into some complications we needed to resolve first. The release is locked in for next Wednesday. Just wanted to thank you for your patience.

**Comment by splix** [comment]
It would be really nice to have it asap. It's a major road blocker right now, since most of the things use Conditional Put now.

**Comment by d--j** [comment]
I just checked support – at least the DE region has conditional PUT support now 🎉

**Comment by kofshikz** [comment]
Yup, can confirm. Just did the Elastic snapshot to the bucket in Frankfurt.

Nice job, guys!

**Comment by lason-ovh** [comment]
Hi all,
As many of you have noticed, we just released the first batch of Conditional Writes this week. It is currently live in the DE region. We are doing a rolling deployment, so it will be available in the remaining regions next week.

**Comment by lason-ovh** [comment]
Hi all,
Release 1 is now officially available on all regions. Release 2 is upcoming in the next weeks.
